CORE-589 Test produce and consume in a multithreaded fashion

ys-pyrofex opened the pull request:
## Overview
The test launches 1000 iterations of one of RSpace examples (Address Book), and shows per-thread and produce/consume ops performance. Example is used as a ground since it better matches real store working conditions.

Also removed some code duplication by refactoring InMemoryStoreTestsBase and LMDBStoreTestsBase

### Does this PR relate to an RChain JIRA issue? 
https://rchain.atlassian.net/browse/CORE-589

lukasz-golebiewski commented on rspace/src/test/scala/coop/rchain/rspace/ConcurrencyTests.scala:
```diff
@@ -0,0 +1,110 @@
+
+          val r3 = produce(store, Channel(channelName), bob, persist = false)
+
+          r3 shouldBe defined
```
I ran those tests on my machine and the assertions are failing:
```
sbt:rspace> testOnly *Concurrency*
[info] Compiling 9 Scala sources to /home/lukasz/git/pyrofex/rchain/rspace/target/scala-2.12/classes ...
[info] Done compiling.
[info] Compiling 1 Scala source to /home/lukasz/git/pyrofex/rchain/rspace/target/scala-2.12/classes ...
[info] Done compiling.
[info] Compiling 16 Scala sources to /home/lukasz/git/pyrofex/rchain/rspace/target/scala-2.12/test-classes ...
[warn] /home/lukasz/git/pyrofex/rchain/rspace/src/test/scala/coop/rchain/rspace/StorageTestsBase.scala:36:58: parameter value sp in class InMemoryStoreTestsBase is never used
[warn]                                                          sp: Serialize[P],
[warn]                                                          ^
[warn] /home/lukasz/git/pyrofex/rchain/rspace/src/test/scala/coop/rchain/rspace/StorageTestsBase.scala:37:58: parameter value sa in class InMemoryStoreTestsBase is never used
[warn]                                                          sa: Serialize[A],
[warn]                                                          ^
[warn] two warnings found
[info] Done compiling.
[info] Compiling 6 Scala sources to /home/lukasz/git/pyrofex/rchain/rspace/target/scala-2.12/test-classes ...
[info] Done compiling.
SLF4J: A number (2) of logging calls during the initialization phase have been intercepted and are
SLF4J: now being replayed. These are subject to the filtering rules of the underlying logging system.
SLF4J: See also http://www.slf4j.org/codes.html#replay
Exception in thread "Thread-21" org.scalatest.exceptions.TestFailedException: None was not defined
        at org.scalatest.MatchersHelper$.indicateFailure(MatchersHelper.scala:346)
        at org.scalatest.Matchers$AnyShouldWrapper.shouldBe(Matchers.scala:7073)
        at coop.rchain.rspace.ConcurrencyTests.$anonfun$$init$$4(ConcurrencyTests.scala:39)
        at coop.rchain.rspace.ConcurrencyTests.$anonfun$$init$$4$adapted(ConcurrencyTests.scala:22)
        at scala.collection.immutable.Range.foreach(Range.scala:156)
        at coop.rchain.rspace.ConcurrencyTests.$anonfun$$init$$3(ConcurrencyTests.scala:22)
        at java.lang.Thread.run(Thread.java:748)
Exception in thread "Thread-22" org.scalatest.exceptions.TestFailedException: None was not defined
        at org.scalatest.MatchersHelper$.indicateFailure(MatchersHelper.scala:346)
        at org.scalatest.Matchers$AnyShouldWrapper.shouldBe(Matchers.scala:7073)
        at coop.rchain.rspace.ConcurrencyTests.$anonfun$$init$$4(ConcurrencyTests.scala:39)
        at coop.rchain.rspace.ConcurrencyTests.$anonfun$$init$$4$adapted(ConcurrencyTests.scala:22)
        at scala.collection.immutable.Range.foreach(Range.scala:156)
        at coop.rchain.rspace.ConcurrencyTests.$anonfun$$init$$3(ConcurrencyTests.scala:22)
        at java.lang.Thread.run(Thread.java:748)
Exception in thread "Thread-27" org.scalatest.exceptions.TestFailedException: None was not defined
        at org.scalatest.MatchersHelper$.indicateFailure(MatchersHelper.scala:346)
        at org.scalatest.Matchers$AnyShouldWrapper.shouldBe(Matchers.scala:7073)
        at coop.rchain.rspace.ConcurrencyTests.$anonfun$$init$$4(ConcurrencyTests.scala:39)
        at coop.rchain.rspace.ConcurrencyTests.$anonfun$$init$$4$adapted(ConcurrencyTests.scala:22)
        at scala.collection.immutable.Range.foreach(Range.scala:156)
        at coop.rchain.rspace.ConcurrencyTests.$anonfun$$init$$3(ConcurrencyTests.scala:22)
        at java.lang.Thread.run(Thread.java:748)
Finished InMemory: 1000 iterations.
```

lukasz-golebiewski replied:
The tests are nevertheless green, because assertion errors were thrown on threads other than the runner thread. I'm not sure if testing correctness and performance should be done in one place. I would tend to use JMH for benchmarks

ys-pyrofex replied:
Hey, thanks for running tests on your machine. We definitely need to test correctness inside multi-threading test - otherwise we will not catch synchronization-related errors.
Seems on your machine the InMemory implementation doesn't work correctly.
What about running the concurrency test on LMDBStore? Is it running without errors or also fails?

lukasz-golebiewski replied:
1) I'm only saying we shouldn't do it in one place, I'd split this into two tests - benchmarks and correctness.
2) `LMDBStore` seems to be working fine

ys-pyrofex replied:
@henrytill - do we need to fix InMemoryStore to pass this test, or just remove it at all, what do you think?

@lukasz-golebiewski sounds reasonable, when I finish merge with latest dev, I will split pure performance test and test+correctness check and compare output. If there will be at least 0.2ms difference - will make them separate.

henrytill commented:
@ys-pyrofex: it's likely going to be easier to deal with exceptions in threads if you use an abstraction like `Task`.

Also makes sense to use `Task` because that's what the Rholang interpreter is using for concurrency.

ys-pyrofex commented:
Added a monix-tasks, implemented unhandled exceptions propagation to main thread for multi-threaded test. Both tests now include warm-up pass for JIT.

henrytill commented:
@ys-pyrofex: could you update?  I think we can move forward now that #926 has been merged.
